sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/DefaultSynchronousMetricStorage.java

```diff
@@ -130,13 +131,15 @@ public MetricData collect(
             : startEpochNanos;
 
     // Grab aggregated points.
-    List<T> points = new ArrayList<>(aggregatorHandles.size());
-    for (Map.Entry<Attributes, AggregatorHandle<T, U>> entry : aggregatorHandles.entrySet()) {
-      T point = entry.getValue().aggregateThenMaybeReset(start, epochNanos, entry.getKey(), reset);
+    points.clear();
+    for (Attributes attributes :
+        ((Map<Attributes, AggregatorHandle<T, U>>) aggregatorHandles).keySet()) {
```

Comment:
I mentioned in the PR description (and in today's spec SIG) that the only remaining allocations are from iterating, like this. Upon closer inspection, it appears that allocations are only necessary while iterating the map if using `map.entrySet()`, which creates `Map.Entry` values.

If we instead iterate over the keyset, we avoid those allocations as well. 

I've updated the PR description and performance figures to reflect this change.

Reply:
What's the difference between this and `forEach` in performance?

Reply:
`forEach` actually performs the same way and was what I ended up going with in #5183.